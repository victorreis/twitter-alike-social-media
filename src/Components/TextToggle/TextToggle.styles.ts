import styled from 'styled-components';

import { Button } from '../Button/Button';

export const TextToggleContainer = styled.div`
  cursor: pointer;
  box-sizing: border-box;
  display: inline-flex;
  align-items: center;

  background-color: ${({ theme }) => theme.colors.background.default.light};

  border-width: 1px;
  border-style: solid;
  border-color: ${({ theme }) => theme.colors.background.default.dark};
  border-radius: 3rem;
`;

export const TextToggleActiveButton = styled(Button)`
  height: 2rem;
  border-radius: 3rem;
  margin: 1px;
  padding: 0 16px;
  border-width: 0;
`;

export const TextToggleButton = styled(TextToggleActiveButton)`
  background-color: ${({ theme }) => theme.colors.background.default.light};
  border-radius: 3rem;

  &:hover {
    background-color: ${({ theme }) => theme.colors.background.default.dark};
  }
`;
