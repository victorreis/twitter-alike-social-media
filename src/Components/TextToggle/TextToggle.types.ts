import { TestProps } from '../../Config/Tests/Test.types';

export interface RequiredTextToggleProps {
  /**
   * Texts to be shown.
   */
  texts: string[];

  /**
   * Callback function that is called when the selected button is changed.
   */
  onToggle: (index: number) => void;
}

export interface DefaultTextToggleProps {
  /**
   * Active index.
   * @default 0
   */
  activeIndex?: number;
}

export interface OptionalTextToggleProps {}

export type TextToggleProps = RequiredTextToggleProps &
  DefaultTextToggleProps &
  OptionalTextToggleProps &
  TestProps &
  Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;
