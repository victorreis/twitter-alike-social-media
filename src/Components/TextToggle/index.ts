export * from './TextToggle.types';
export * from './TextToggle';
