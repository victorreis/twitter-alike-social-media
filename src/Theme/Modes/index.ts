export { darkTheme } from './Dark.theme';
