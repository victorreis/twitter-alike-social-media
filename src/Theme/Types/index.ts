export * from './Borders.types';
export * from './Colors.types';
export * from './CustomTheme.types';
export * from './Sizes.types';
export * from './Typographies.types';
