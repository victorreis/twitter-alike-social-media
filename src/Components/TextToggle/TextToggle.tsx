import { useState, useEffect } from 'react';

import { nanoid } from 'nanoid';

import { TestProps } from '../../Config/Tests/Test.types';
import {
  TextToggleButton,
  TextToggleActiveButton,
  TextToggleContainer,
} from './TextToggle.styles';
import { TextToggleProps, DefaultTextToggleProps } from './TextToggle.types';

export const textToggleDefaults: Required<DefaultTextToggleProps> &
  Required<TestProps> = {
  testID: 'TextToggle',
  activeIndex: 0,
};

export const TextToggle: React.FC<TextToggleProps> = (props): JSX.Element => {
  const {
    testID = textToggleDefaults.testID,
    texts,
    onToggle,
    activeIndex: defaultIndex = textToggleDefaults.activeIndex,
    style,
    ...others
  } = props;

  const [activeIndex, setActiveIndex] = useState<number>(defaultIndex);

  useEffect(() => {
    setActiveIndex(() => defaultIndex);
  }, [defaultIndex]);

  const renderButtons = () => {
    return texts.map((text, index) => {
      const handleClick = () => {
        setActiveIndex(index);
        onToggle(index);
      };

      if (activeIndex === index) {
        return (
          <TextToggleActiveButton key={nanoid()} onClick={() => {}} size="MD">
            {text}
          </TextToggleActiveButton>
        );
      }
      return (
        <TextToggleButton key={nanoid()} onClick={handleClick} size="MD">
          {text}
        </TextToggleButton>
      );
    });
  };

  return (
    <TextToggleContainer data-testid={testID} style={style} {...others}>
      {renderButtons()}
    </TextToggleContainer>
  );
};
