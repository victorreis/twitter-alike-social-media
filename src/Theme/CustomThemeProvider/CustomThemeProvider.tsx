/* eslint-disable @typescript-eslint/ban-ts-comment */
import { useMemo, useState, createContext } from 'react';
import { BrowserRouter } from 'react-router-dom';

import { ThemeProvider } from 'styled-components';

import { darkTheme } from '../Modes';
import { ThemeMode, CustomThemeType } from '../Types';
import {
  ThemeContextType,
  CustomThemeProviderProps,
  DefaultCustomThemeProviderProps,
} from './CustomThemeProvider.types';

export const themes: Record<ThemeMode | 'default', CustomThemeType> = {
  default: darkTheme,
  dark: darkTheme,
};

export const ThemeContext = createContext<
  ThemeContextType | Record<string, never>
>({});

export const customThemeProviderDefaults: Required<DefaultCustomThemeProviderProps> =
  {
    testID: 'ThemeContextProvider',
    themeName: 'dark',
  };

export const CustomThemeProvider: React.FC<CustomThemeProviderProps> = (
  props
): JSX.Element => {
  const {
    testID = customThemeProviderDefaults.testID,
    children,
    themeName = customThemeProviderDefaults.themeName,
  } = props;

  const [theme, setTheme] = useState<CustomThemeType>(themes[themeName]);

  const switchTheme = (themeMode: ThemeMode): void => {
    setTheme(themes[themeMode]);
  };

  const providerValue = useMemo(() => ({ theme, switchTheme }), [theme]);

  return (
    <ThemeContext.Provider data-testid={testID} value={providerValue}>
      {/* waiting for a new lib version compatible with React 18 */}
      {/* @ts-ignore */}
      <ThemeProvider theme={theme}>
        <BrowserRouter>{children}</BrowserRouter>
      </ThemeProvider>
    </ThemeContext.Provider>
  );
};
