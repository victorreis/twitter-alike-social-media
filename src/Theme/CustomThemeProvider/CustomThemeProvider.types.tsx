import { CustomThemeType, ThemeMode } from '../Types';

export type ThemeContextType = {
  theme: CustomThemeType;
  switchTheme: (themeMode: ThemeMode) => void;
};

export interface RequiredCustomThemeProviderProps {
  children?: React.ReactNode;
}

export interface DefaultCustomThemeProviderProps {
  testID?: string;
  themeName?: ThemeMode;
}

export type CustomThemeProviderProps = RequiredCustomThemeProviderProps &
  DefaultCustomThemeProviderProps;
