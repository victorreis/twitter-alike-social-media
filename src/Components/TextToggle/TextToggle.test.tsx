import {
  fireEvent,
  renderJestDomCreator,
  renderRTRCreator,
  screen,
} from '../../Config/Tests/GlobalSetup.config';
import { themes } from '../../Theme/CustomThemeProvider';
import { hexToRgb } from '../../Utils/Transform/hexToRgb.util';
import { TextToggle, textToggleDefaults } from './TextToggle';
import { RequiredTextToggleProps, TextToggleProps } from './TextToggle.types';

describe('textToggle component tests', () => {
  const texts = ['teste1', 'teste2', 'teste3'];
  const onToggle = jest.fn();

  const requiredProps: RequiredTextToggleProps = {
    texts,
    onToggle,
  };

  const setup = (props?: TextToggleProps) => {
    const renderRTR = () =>
      renderRTRCreator<TextToggleProps>(TextToggle, {
        ...requiredProps,
        ...props,
      });
    const renderJestDom = () =>
      renderJestDomCreator<TextToggleProps>(TextToggle, {
        ...requiredProps,
        ...props,
      });

    return { renderRTR, renderJestDom };
  };

  describe('behavior tests', () => {
    it(`should render the component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const testInstance = screen.getByTestId(textToggleDefaults.testID);

      expect(testInstance).toBeTruthy();
    });

    it(`should render the texts`, () => {
      expect.assertions(3);
      setup().renderJestDom();
      const element0 = screen.getByText(String(texts[0]));
      const element1 = screen.getByText(String(texts[1]));
      const element2 = screen.getByText(String(texts[2]));

      expect(element0).toBeInTheDocument();
      expect(element1).toBeInTheDocument();
      expect(element2).toBeInTheDocument();
    });

    it(`should call the 'onToggle' callback when the button is clicked`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const index = 1;
      const element = screen.getByText(String(texts[index]));

      fireEvent.click(element);

      expect(onToggle).toHaveBeenCalledWith(index);
    });
  });

  describe('style tests', () => {
    it(`should have styled the Container component`, () => {
      expect.assertions(1);
      setup().renderJestDom();
      const container = screen.getByTestId(textToggleDefaults.testID);

      expect(container).toHaveStyle({
        backgroundColor: themes.default.colors.background.default.light,
        borderColor: themes.default.colors.background.default.dark,
      });
    });

    it(`should change style of the buttons when some button is clicked`, () => {
      expect.assertions(4);
      setup().renderJestDom();
      const elementActiveByDefault = screen.getByText(
        String(texts[textToggleDefaults.activeIndex])
      );
      const element2 = screen.getByText(String(texts[1]));

      expect(elementActiveByDefault).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.darkest
        ),
      });
      expect(element2).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.light
        ),
      });

      fireEvent.click(element2);

      // TODO: async changes
      expect(elementActiveByDefault).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.darkest
        ),
      });
      expect(element2).toHaveStyle({
        backgroundColor: hexToRgb(
          themes.default.colors.background.default.light
        ),
      });
    });
  });

  describe('snapshot tests', () => {
    it(`should render correctly`, () => {
      expect.assertions(1);
      const generatedJson = setup().renderRTR().toJSON();

      expect(generatedJson).toMatchSnapshot();
    });
  });
});
