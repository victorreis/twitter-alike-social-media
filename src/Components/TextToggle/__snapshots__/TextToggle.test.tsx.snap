// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`textToggle component tests snapshot tests should render correctly 1`] = `
<div
  className="TextToggleContainer-twitter-alike-social-media__sc-oz9uiz-0 isqEII"
  data-testid="TextToggle"
>
  <button
    className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 dLbeGW TextToggleActiveButton-twitter-alike-social-media__sc-oz9uiz-1 jwXQve"
    data-testid="Button"
    disabled={false}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size="MD"
    type="button"
  >
    teste1
  </button>
  <button
    className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 dLbeGW TextToggleActiveButton-twitter-alike-social-media__sc-oz9uiz-1 TextToggleButton-twitter-alike-social-media__sc-oz9uiz-2 jwXQve eUAzmR"
    data-testid="Button"
    disabled={false}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size="MD"
    type="button"
  >
    teste2
  </button>
  <button
    className="ButtonContainer-twitter-alike-social-media__sc-1bqjegs-0 dLbeGW TextToggleActiveButton-twitter-alike-social-media__sc-oz9uiz-1 TextToggleButton-twitter-alike-social-media__sc-oz9uiz-2 jwXQve eUAzmR"
    data-testid="Button"
    disabled={false}
    onClick={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    size="MD"
    type="button"
  >
    teste3
  </button>
</div>
`;
